File: packages/moss-repo-indexer/src/moss_repo_indexer/clone.py
"""Resolve a local path or shallow-clone a git URL into a working directory."""

from __future__ import annotations

import re
import shutil
import subprocess
import tempfile
from dataclasses import dataclass
from pathlib import Path
from typing import Optional
from urllib.parse import urlparse

_GIT_SSH_RE = re.compile(r"^git@[^:]+:.+\.git$")
_GIT_URL_RE = re.compile(r"^(https?|git)://", re.IGNORECASE)


@dataclass
class ResolvedSource:
    root: Path
    repo_name: str
    ref: Optional[str]
    cleanup: bool

    def close(self) -> None:
        if self.cleanup and self.root.exists():
            shutil.rmtree(self.root, ignore_errors=True)


def resolve_source(source: str, ref: Optional[str] = None) -> ResolvedSource:
    """Return a local directory for `source` (filesystem path or git URL)."""
    stripped = source.strip()
    if _is_git_url(stripped):
        return _clone_git_url(stripped, ref)
    return _resolve_local_path(stripped, ref)


def _is_git_url(source: str) -> bool:
    if _GIT_SSH_RE.match(source) or _GIT_URL_RE.match(source):
        return True
    if source.endswith(".git") and "://" in source:
        return True
    return False


def _resolve_local_path(source: str, ref: Optional[str]) -> ResolvedSource:
    root = Path(source).expanduser().resolve()
    if not root.is_dir():
        raise FileNotFoundError(f"Source path is not a directory: {root}")
    return ResolvedSource(
        root=root,
        repo_name=root.name,
        ref=ref,
        cleanup=False,
    )


def _clone_git_url(url: str, ref: Optional[str]) -> ResolvedSource:
    if shutil.which("git") is None:
        raise RuntimeError("git is required to clone a repository URL")

    temp_dir = Path(tempfile.mkdtemp(prefix="moss-repo-"))
    try:
        _run_git_clone(url, temp_dir, ref)
    except Exception:
        shutil.rmtree(temp_dir, ignore_errors=True)
        raise

    return ResolvedSource(
        root=temp_dir,
        repo_name=_repo_name_from_url(url),
        ref=ref,
        cleanup=True,
    )


def _run_git_clone(url: str, dest: Path, ref: Optional[str]) -> None:
    command = ["git", "clone", "--depth", "1"]
    if ref:
        command.extend(["--branch", ref])
    command.extend([url, str(dest)])
    result = subprocess.run(command, capture_output=True, text=True, check=False)
    if result.returncode != 0:
        detail = (result.stderr or result.stdout or "git clone failed").strip()
        raise RuntimeError(f"Failed to clone {url}: {detail}")


def _repo_name_from_url(url: str) -> str:
    if _GIT_SSH_RE.match(url):
        name = url.rsplit(":", 1)[-1]
    else:
        name = urlparse(url).path
    name = Path(name).name
    if name.endswith(".git"):
        name = name[:-4]
    return name or "repo"

File: packages/moss-repo-indexer-js/src/clone.ts
import { spawnSync } from 'node:child_process'
import fs from 'node:fs'
import os from 'node:os'
import path from 'node:path'

const GIT_SSH_RE = /^git@[^:]+:.+\.git$/
const GIT_URL_RE = /^(https?|git):\/\//i

export interface ResolvedSource {
  root: string
  repoName: string
  ref?: string
  cleanup: boolean
  close: () => void
}

export function resolveSource(source: string, ref?: string): ResolvedSource {
  const stripped = source.trim()
  if (isGitUrl(stripped)) {
    return cloneGitUrl(stripped, ref)
  }
  return resolveLocalPath(stripped, ref)
}

export function isGitUrl(source: string): boolean {
  if (GIT_SSH_RE.test(source) || GIT_URL_RE.test(source)) {
    return true
  }
  return source.endsWith('.git') && source.includes('://')
}

export function repoNameFromUrl(url: string): string {
  let name = GIT_SSH_RE.test(url) ? url.split(':').pop() ?? url : new URL(url).pathname
  name = path.basename(name)
  if (name.endsWith('.git')) {
    name = name.slice(0, -4)
  }
  return name || 'repo'
}

function resolveLocalPath(source: string, ref?: string): ResolvedSource {
  const root = path.resolve(source)
  if (!fs.existsSync(root) || !fs.statSync(root).isDirectory()) {
    throw new Error(`Source path is not a directory: ${root}`)
  }
  return {
    root,
    repoName: path.basename(root),
    ref,
    cleanup: false,
    close: () => undefined,
  }
}

function cloneGitUrl(url: string, ref?: string): ResolvedSource {
  if (!hasGit()) {
    throw new Error('git is required to clone a repository URL')
  }
  const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'moss-repo-'))
  try {
    runGitClone(url, tempDir, ref)
  } catch (error) {
    fs.rmSync(tempDir, { recursive: true, force: true })
    throw error
  }
  return {
    root: tempDir,
    repoName: repoNameFromUrl(url),
    ref,
    cleanup: true,
    close: () => {
      fs.rmSync(tempDir, { recursive: true, force: true })
    },
  }
}

function hasGit(): boolean {
  return spawnSync('git', ['--version'], { encoding: 'utf8' }).status === 0
}

function runGitClone(url: string, dest: string, ref?: string): void {
  const args = ['clone', '--depth', '1']
  if (ref) {
    args.push('--branch', ref)
  }
  args.push(url, dest)
  const result = spawnSync('git', args, { encoding: 'utf8' })
  if (result.status !== 0) {
    const detail = (result.stderr || result.stdout || 'git clone failed').trim()
    throw new Error(`Failed to clone ${url}: ${detail}`)
  }
}
